python/hail/methods/statgen.py

```diff
@@ -2908,15 +2909,103 @@ def filter(self):
+    """Prune variants in linkage disequilibrium.
+
+    Notes
+    -----
+
+    This method prunes variants in linkage disequilibrium in two stages. The first stage
+    is a local pruning step, which prunes variants in the same partition. The parallelism
+    in this step will be affected by the number of partitions in the dataset, as well as
+    the memory per core, which is used to calculate a queue size for the local prune.
+
+    The second stage is a global pruning step which makes use of a correlation matrix
+    across all remaining variants. In the global pruning step, correlated variants are
+    passed to a method that computes a maximal independent set of those variants.
+
+    Parameters
+    ----------
+    ds : :class:`.MatrixTable`
+        dataset to prune
+    r2 : :obj:`float`
+        correlation threshold (exclusive) above which variants are removed
+    window : :obj:`int`
+        distance in kilobases; correlation between variants further apart is not considered in pruning
+    memory_per_core : :obj:`int`
+        memory in MB per core for local pruning
+
+    Returns
+    -------
+    :class:`.Table`
+        Table of variants after pruning.
+    """
+    bytes_per_core = memory_per_core * 1024 * 1024
+    fraction_memory_to_use = 0.25
+    variant_byte_overhead = 50
+    genotypes_per_pack = 32
+    n_samples = ds.count_cols()
+    min_memory_per_core = math.ceil((1 / fraction_memory_to_use) * 8 * n_samples + variant_byte_overhead)
+    if bytes_per_core < min_memory_per_core:
+        raise ValueError("memory per core must be greater than {} MB".format(min_memory_per_core * 1024 * 1024))
+    bytes_per_variant = math.ceil(8 * n_samples / genotypes_per_pack) + variant_byte_overhead
+    memory_available_per_core = bytes_per_core * fraction_memory_to_use
+    max_queue_size = int(max(1.0, math.ceil(memory_available_per_core / bytes_per_variant)))
+
+    sites_only_table = Table(Env.hail().methods.LocalLDPrune.apply(
+        require_biallelic(ds, 'ld_prune')._jvds, float(r2), window, max_queue_size))
+
+    sites_path = new_temp_file()
+    sites_only_table.write(sites_path, overwrite=True)
+    sites_only_table = hl.read_table(sites_path)
+
+    locally_pruned_ds = ds.annotate_rows(pruned=sites_only_table[ds.row_key])
+    locally_pruned_ds = locally_pruned_ds.filter_rows(hl.is_defined(locally_pruned_ds.pruned))
+
+    locally_pruned_ds = (locally_pruned_ds.annotate_rows(
+        mean=locally_pruned_ds.pruned.mean, sd_reciprocal=locally_pruned_ds.pruned.sd_reciprocal)
+        .add_row_index('row_idx'))
+
+    locally_pruned_path = new_temp_file()
+    locally_pruned_ds.write(locally_pruned_path, overwrite=True)
+    locally_pruned_ds = hl.read_matrix_table(locally_pruned_path)
+
+    normalized_mean_imputed_genotype_expr = (
+        hl.cond(hl.is_defined(locally_pruned_ds.GT),
+                (locally_pruned_ds.GT.n_alt_alleles() - locally_pruned_ds.mean)
+                * locally_pruned_ds.sd_reciprocal, 0))
```

Comment:
I'm a bit confused here: if  `sd_reciprocal` is really the reciprocal of standard deviation, then the resulting rows should have mean 0 and standard deviation 1, so vector length `n`. But the `correlation_matrix` defined as `block_matrix @ block_matrix.T` is then going to have `n` on the diagonal rather than `1`.

So either `sd_reciprocal` is really `1 / length = 1 / (n * sd)` and is poorly named, or the correlation matrix is off by a factor of `n`.

Reply:
Er, yes, I was looking at this before and sd_reciprocal is 1/ (sqrt(n) * sd), which I got from the local pruning step. I don't think the standard deviation is inside the square root like you've written? The correlation matrix has 1's along the diagonal, and I checked that the standard deviations match those given by agg.stats when divided by sqrt(n). I will either change the variable names or reformulate the code to clarify this point.

Reply:
Great. Yup, i meant to write `sd` outside `sqrt`, fixed. I think `centered_length_reciprocal` is accurate.